Updated member email alerts to trigger via events

rshbhgrg opened the pull request:
closes https://github.com/TryGhost/Team/issues/1864
closes https://github.com/TryGhost/Team/issues/1865
closes https://github.com/TryGhost/Team/issues/1867
refs https://github.com/TryGhost/Team/issues/1881

- updates sending member alert emails using events instead of adding a direct dependency which is harder to maintain long term
- using events also allows decoupling the actual email sending from the main request for an action like cancel subscription or paid subscription

rishabhgrg commented on ghost/member-events/lib/PaidSubscriptionStartedEvent.js:
```diff
@@ -0,0 +1,60 @@
+/**
```
Had 2 minds about overloading the existing SubscriptionCreatedEvent vs adding a new one which captures all the information for paid susbcription, and a new one seemed better as the two have different usecases.

allouis replied:
I think using the `SubscriptionCreatedEvent` makes more sense here ☺️ We have an issue to move it here: https://github.com/TryGhost/Team/issues/1881

So would be great if you could use that instead

rishabhgrg replied:
Ah! So we'd want the `SubscriptionCreatedEvent` to be moved to where we have the new `PaidSubscriptionStartedEvent` in `linkSubscription` right for that issue? Guess that'll also need to pass down the attribution changes ? And how do you feel about overloading it with full data instead of IDs ?

rishabhgrg replied:
Resolved below, gonna switch to using ids

rishabhgrg commented on ghost/member-events/lib/MemberSubscriptionCancellationEvent.js:
```diff
@@ -0,0 +1,52 @@
+/**
```
Noticed while adding this, we also have [MemberPaidCancellationEvent](https://github.com/TryGhost/Ghost/blob/7084217d3df39fbf6f3cba302c63414282882873/ghost/member-events/lib/MemberPaidCancellationEvent.js) which was added for Analytics but doesn't seem to have any active usecases. Maybe a good candidate for removing in future to avoid having two similar named events

allouis replied:
How do you feel about `SubscriptionCancelledEvent` to match with `SubscriptionCreatedEvent`?

rishabhgrg replied:
Sounds good if we are moving to `SubscriptionCreatedEvent` above :)

allouis commented on ghost/members-api/lib/repositories/member.js:
```diff
@@ -967,15 +993,30 @@ module.exports = class MemberRepository {
+                duration: offer.duration.value
+            } : null;
+
+            DomainEvents.dispatch(PaidSubscriptionStartedEvent.create({
```
If we switch to use the `SubscriptionCreatedEvent` you'll notice that we just pass id's through - which I know is a bit of a PITA, but I think it's the best way to go until we've really nailed down this event pattern.

rishabhgrg replied:
My concern with passing IDs only is that it's not just one data object that we are fetching, we'll need to make requests to fetch multiple bits like the associated offer/tier/member other than the subscription, which is a lot of overhead for values that we already have 🤔

rishabhgrg replied:
Gonna switch to using IDs for now so we can come back and clean this up later with a proper data pattern

allouis commented on ghost/members-api/lib/repositories/member.js:
```diff
@@ -1232,21 +1273,37 @@ module.exports = class MemberRepository {
+                const context = options?.context || {};
+                const source = this._resolveContextSource(context);
+
+                DomainEvents.dispatch(MemberSubscriptionCancellationEvent.create({
```
As above - do you mind if we just pass ids for now? I'm concerned about having different data shapes used throughout the app, this event pattern is still fairly new and hasn't quite settled yet.

rishabhgrg replied:
👍

allouis reviewed (commented):
This approach looks so much better!

Few requests though:
- Reuse the existing `SubscriptionCreatedEvent`
- Pass id's in event data for now
- Rename `PaidSubscriptionCanceledEvent` to `SubscriptionCancelledEvent` for consistency with the created event

rshbhgrg committed:
Updated staff service to trigger alerts via events

refs https://github.com/TryGhost/Team/issues/1865

- refactors staff service to listen to member and subscription events
- triggers email alerts based on events instead of directly calling the service
- removes staff service dependency for members api

rshbhgrg committed:
Updated subscription created/canceled events data

refs https://github.com/TryGhost/Team/issues/1865

- adds new subscription canceled event
- updates existing subscription created event to include tier id and source data

rshbhgrg committed:
Moved member email alert trigger to member creation

closes https://github.com/TryGhost/Team/issues/1864
refs https://github.com/TryGhost/Team/issues/1881

- triggers free member email alert via event dispatch from member create method
- passes subscription/stripe data to member creation for paid members so free member alert can be ignored for them
- moves subscription created event being called from webhook controller to `linkSubscription`, allows creating subscription events for all new subscriptions instead of ones just via webhooks

rshbhgrg committed:
Wired events for triggering email alerts for subscription creation/cancellation

refs https://github.com/TryGhost/Team/issues/1865

- refactors subscription creation/cancellation to dispatch proper events which are used for email alerts
- cleanup
